File: src/com/kraftwerking/datastructuresalgo/leetcode/LongestConsecutiveSubsequence128.java
package com.kraftwerking.datastructuresalgo.leetcode;


import java.util.HashSet;

class LongestConsecutiveSubsequence128 {

  public int longestConsecutive(int[] nums) {
    HashSet<Integer> numSet = new HashSet<>();
    for (Integer n : nums) {
      numSet.add(n);
    }
    int longest = 0;

    for(int i = 0;i<nums.length;i++){
      if(!numSet.contains(nums[i] - 1)){
        int len = 0;
        int curr = nums[i];

        while(numSet.contains(curr)){
          curr = curr + 1;
          longest = Math.max(longest, ++len);
        }

      }
    }

    return longest;
  }

  public static void main(String[] args) {
    int nums[] = {0,3,7,2,5,8,4,6,0,1};
    LongestConsecutiveSubsequence128 longestConsecutiveSubsequence128 = new LongestConsecutiveSubsequence128();
    int result = longestConsecutiveSubsequence128.longestConsecutive(nums);
    System.out.println(result);
  }
}

File: python/leetcode/LongestConsecutiveSubsequence128.py
from typing import List


class Solution(object):
    def longestConsecutive(self, nums: List[int]) -> int:
        numSet = set(nums)
        longest = 0

        for n in nums:
            # check if its the start of a sequence
            if(n - 1) not in numSet:
                length = 0
                while(n + length) in numSet:
                    length += 1
                longest = max(length, longest)
        return longest


def main():
    nums = [0, 3, 7, 2, 5, 8, 4, 6, 0, 1]

    solution = Solution()
    result = solution.longestConsecutive(nums)
    print(result)


if __name__ == "__main__":
    main()
